src/rgw/rgw_rest_user_policy.cc

```diff
@@ -38,377 +37,673 @@ void RGWRestUserPolicy::send_response()
+{
+}
+
+int RGWListUserPolicies::get_params()
+{
+  marker = s->info.args.get("Marker");
+
+  int r = s->info.args.get_int("MaxItems", &max_items, max_items);
+  if (r < 0 || max_items > 1000) {
+    s->err.message = "Invalid value for MaxItems";
+    return -EINVAL;
+  }
+
+  return RGWRestUserPolicy::get_params();
+}
+
+void RGWListUserPolicies::execute(optional_yield y)
+{
+  std::map<std::string, std::string> policies;
+  if (auto it = user->get_attrs().find(RGW_ATTR_USER_POLICY); it != user->get_attrs().end()) {
+    try {
+      decode(policies, it->second);
+    } catch (buffer::error& err) {
+      ldpp_dout(this, 0) << "ERROR: failed to decode user policies" << dendl;
+      op_ret = -EIO;
       return;
     }
-    s->formatter->close_section();
-    s->formatter->close_section();
   }
-  if (op_ret < 0) {
-    op_ret = -ERR_INTERNAL_ERROR;
+
+  s->formatter->open_object_section_in_ns("ListUserPoliciesResponse", RGW_REST_IAM_XMLNS);
+  s->formatter->open_object_section("ResponseMetadata");
+  s->formatter->dump_string("RequestId", s->trans_id);
+  s->formatter->close_section();
+  s->formatter->open_object_section("ListUserPoliciesResult");
+  s->formatter->open_array_section("PolicyNames");
+  auto policy = policies.lower_bound(marker);
+  for (; policy != policies.end() && max_items > 0; ++policy, --max_items) {
+    s->formatter->dump_string("member", policy->first);
+  }
+  s->formatter->close_section(); // PolicyNames
+  const bool is_truncated = (policy != policies.end());
+  encode_json("IsTruncated", is_truncated, s->formatter);
+  if (is_truncated) {
+    encode_json("Marker", policy->first, s->formatter);
   }
+  s->formatter->close_section(); // ListUserPoliciesResult
+  s->formatter->close_section(); // ListUserPoliciesResponse
 }
 
-uint64_t RGWListUserPolicies::get_op()
+
+RGWDeleteUserPolicy::RGWDeleteUserPolicy(const ceph::bufferlist& post_body)
+  : RGWRestUserPolicy(rgw::IAM::iamDeleteUserPolicy, RGW_CAP_WRITE),
+    post_body(post_body)
 {
-  return rgw::IAM::iamListUserPolicies;
 }
 
-int RGWListUserPolicies::get_params()
+int RGWDeleteUserPolicy::get_params()
 {
-  user_name = s->info.args.get("UserName");
+  policy_name = s->info.args.get("PolicyName");
+  if (!validate_iam_policy_name(policy_name, s->err.message)) {
+    return -EINVAL;
+  }
 
-  if (user_name.empty()) {
-    ldpp_dout(this, 20) << "ERROR: user name is empty" << dendl;
+  return RGWRestUserPolicy::get_params();
+}
+
+int RGWDeleteUserPolicy::forward_to_master(optional_yield y, const rgw::SiteConfig& site)
+{
+  RGWXMLDecoder::XMLParser parser;
+  if (!parser.init()) {
+    ldpp_dout(this, 0) << "ERROR: failed to initialize xml parser" << dendl;
     return -EINVAL;
   }
 
+  s->info.args.remove("UserName");
+  s->info.args.remove("PolicyName");
+  s->info.args.remove("Action");
+  s->info.args.remove("Version");
+
+  int r = forward_iam_request_to_master(this, site, s->user->get_info(),
```

Comment:
@cbodley : why do we explicitly need to send these requests for replication, won't the policies replicate as part of the user? or are these required for incremental sync?

Reply:
What I am trying to ask is that, isnt this `forward_iam_request_to_master` taken care of in svc_user_rados.cc, where the user metadata is replicated in `PutOperation`.

Reply:
when a secondary zone replicates user metadata from the master zone, it does call through RGWUserMetadataHandler::do_put() -> store_user_info() -> PutOperation

but if we write new metadata to a non-master zone, no other zones will try to replicate that. other zones only run metadata sync against the metadata master zone

forwarding metadata write requests to the master zone is an important part of our metadata consistency model

Reply:
thank you for explaining this.